syntax = "proto3";

package mir.command;

/// assertion type: training, validation or test
enum TvtType {
    TvtTypeUnknown = 0;
    TvtTypeTraining = 1;
    TvtTypeValidation = 2;
    TvtTypeTest = 3;
};

enum AssetType {
    AssetTypeUnknown = 0;
    AssetTypeImageJpeg = 1;
    AssetTypeImagePng = 2;
    AssetTypeImagePixelMat = 3;
    AssetTypeImageYuv420p = 4;
    AssetTypeImageYuv420sp = 5;
    AssetTypeImageYuv422p = 6;
    AssetTypeImageYuv422sp = 7;
    AssetTypeImageBmp = 8;
    AssetTypeVideoMp4 = 101;
};

/// task type
enum TaskType {
    TaskTypeUnknown = 0;
    TaskTypeTraining = 1;
    TaskTypeMining = 2;
    TaskTypeLabel = 3;
    TaskTypeFilter = 4;
    TaskTypeImportData = 5;
    TaskTypeExportData = 6;
    TaskTypeCopyData = 7;
    TaskTypeMerge = 8;
};

enum TaskState {
    TaskStateUnknown = 0;
    TaskStatePending = 1;
    TaskStateRunning = 2;
    TaskStateDone = 3;
    TaskStateError = 4;
    TaskStateMiss = 5;
};

enum Sha1Type {
    SHA1_TYPE_UNKNOWN = 0;
    SHA1_TYPE_ASSET = 1;
    SHA1_TYPE_COMMIT = 2;
}

enum RCode {
    RC_OK = 0;
    RC_ERROR_UNKNOWN = 1;

    RC_CMD_ERROR_UNKNOWN = 100;
    RC_CMD_CORRUPT_MIR_FILE = 101;
    RC_CMD_INVALID_MIR_FILE = 102;
    RC_CMD_INVALID_MIR_REPO = 103;
    RC_CMD_INVALID_ARGS = 104;
    RC_CMD_INVALID_BRANCH_OR_TAG = 105;
    RC_CMD_INVALID_COMMAND = 106;
    RC_CMD_MIR_FILE_NOT_FOUND = 107;
    RC_CMD_CONFLICTS_OCCURED = 108;
    RC_CMD_EMPTY_METADATAS = 109;
    RC_CMD_EMPTY_TRAIN_SET = 110;
    RC_CMD_EMPTY_VAL_SET = 111;
    RC_CMD_DIRTY_REPO = 112;
    RC_CMD_NOTHING_TO_MERGE = 113;

    RC_SERVICE_ERROR_UNKNOWN = 1000;
    RC_SERVICE_INVALID_REQ = 1001;
    RC_SERVICE_INVOKE_ERROR = 1002;
    RC_SERVICE_UNKOWN_RESPONSE = 1003;
    RC_SERVICE_TASK_INVOKER_ERROR = 1004;
    RC_SERVICE_INFO_INVOKER_ERROR = 1005;
}

enum MirStorage {
    MIR_METADATAS = 0;
    MIR_ANNOTATIONS = 1;
    MIR_KEYWORDS = 2;
    MIR_TASKS = 3;
}

enum LabelFormat {
    NO_ANNOTATION = 0;
    PASCAL_VOC = 1;
    IF_ARK = 2;
};

/// ========== metadatas.mir ==========
message MirMetadatas {
    /// key: asset hash, value: attributes
    map<string, MetadataAttributes> attributes = 1;
};

message MetadataAttributes {
    string dataset_name = 1;
    Timestamp timestamp = 2;
    TvtType tvt_type = 3;
    AssetType asset_type = 4;
    int32 width = 5;  /// column number
    int32 height = 6;  /// row number
    int32 image_channels = 7;  /// (for images) channel count
};

message Timestamp {
    /// start time stamp
    int64 start = 1;
    /// duration (in seconds), for images, it's always 0
    float duration = 2;
};

/// ========== annotations.mir ==========
message MirAnnotations {
    /// key: task id, value: annotations of that single task
    map<string, SingleTaskAnnotations> task_annotations = 1;
    string head_task_id = 2;
};

message SingleTaskAnnotations {
    /// key: image id, value: annotations of that single image
    map<string, SingleImageAnnotations> image_annotations = 1;
};

message SingleImageAnnotations {
    repeated Annotation annotations = 2;
};

message Annotation {
    // Index of this annotation in current single image, may be different from the index in repeated field.
    int32 index = 1;
    Rect box = 2;
    int32 class_id = 3;
    double score = 4;
};

message Rect {
    int32 x = 1;
    int32 y = 2;
    int32 w = 3;
    int32 h = 4;
};

/// ========== keywords.mir ==========
message MirKeywords {
    // key: asset hash, value: keywords list
    // cnt: count of keywords
    map<string, Keywords> keywords = 1;

    map<int32, int32> predifined_keyids_cnt = 2;
    // sum of predifined_keyids_cnt field.
    int32 predifined_keyids_total = 3;

    map<string, int32> customized_keywords_cnt = 4;
    // sum of predifined_keyids_cnt field.
    int32 customized_keywords_total = 5;

    // store index map from key_id to set of asset_ids.
    map<int32, Assets> index_predifined_keyids = 6;
};

message Assets {
    repeated string asset_ids = 1;
};

message Keywords {
    // predefined: managed id-keyword map
    repeated int32 predifined_keyids = 1;
    // customized: arbitrary user defined keywords
    repeated string customized_keywords = 2;
};

/// ========== tasks.mir ==========
message MirTasks {
    map<string, Task> tasks = 1;
    string head_task_id = 2;
};

message Task {
    TaskType type = 1;
    /// user defined task name
    string name = 2;
    /// auto generated unique id
    string task_id = 3;
    /// auto generated unique id
    string base_task_id = 4;
    /// execution time of this task
    int64 timestamp = 5;
    ModelMeta model = 6;
    map<string, int32> unknown_types = 7;
    /// serialization result of arguments of this task, in yaml format
    string args = 100;
    string ancestor_task_id = 101;
};

message ModelMeta {
    /// hash for models.tar.gz
    string model_hash = 1;
    /// model mAP
    float mean_average_precision = 2;
    /// context generated by train command
    string context = 3;
};
